### DISCUSSION TITLE
"normal" field for terms no longer exists in v13?

### DISCUSSION THREAD

**Discussion by scpedicini** [nlp-compromise/nlp_compromise#854] [Q&A] [answered]
I used to use v11 heavily in a project a long time ago, and have been recently going through and updating it to use it with v13 compromise but it's missing some older properties/fields. I used to break apart each "compound verb" using v.list[0].terms and iterating through each term but the "normal" field is missing. Has "normal" been renamed? I noticed other fields 'reduced' and 'cleaned'. Can I substitute normal for one of these?

Also, I remember that 'silent_term' was related to contractions, but can't find mention of that field either.

**Comment by spencermountain** [comment, marked as answer]
hey @scpedicini yeah the internal api changes and will change again in v14. You should be able to break apart a match into individual terms with `.terms()` and get request any metadata, such as normalized, or implict words, with [.json()](https://observablehq.com/@spencermountain/compromise-json)
cheers

**Reply by scpedicini** [reply]
Hey Spencer, thanks for the reply. I used to gravitate towards the terms/tag values because it seemed a good way to analyze POS non-destructively. 

So would something like this be best practice?
`json_out = nlp('They broke up' ).verbs(0).terms().json();`
`broke_tags = json_out[0].terms[0].tags;`

OR should I use the list[0]:
`nlp('They broke up' ).verbs(0).list[0].terms()`

**Reply by spencermountain** [reply]
yeah, the former one. The (rough) idea is that `.json()` results can be whipped around with javascript, changed however you'd like, and then thrown back in with `.import()`

admittedly, this isn't working that well, and I'm open to suggestions for improving it. `.list[]` is being canned in the next version, which should land in August.

**Reply by spencermountain** [reply]
any results of .json() will be non-destructive, or won't actually change the document.

**Reply by scpedicini** [reply]
Hey Spencer,

Thanks again, jjust to confirm (this time with a noun) - these produce relatively similar JSONified output, but I think the second command nouns(0).json()[0] is the one I should use. Does that look right in terms of future-proofing myself for the upcoming v14?

_I think this one is not best practice_
`nlp('A Great Dane').nouns(0).terms().json()`
```
[
  {
    "text": "Great",
    "terms": [
      {
        "text": "Great",
        "tags": [
          "ProperNoun",
          "Noun",
          "Singular",
          "Person"
        ],
        "pre": "",
        "post": " "
      }
    ]
  },
  {
    "text": "Dane",
    "terms": [
      {
        "text": "Dane",
        "tags": [
          "MaleName",
          "FirstName",
          "Person",
          "ProperNoun",
          "Singular",
          "Noun"
        ],
        "pre": "",
        "post": ""
      }
    ]
  }
]
```

_This one is right? and then I can just iterate through the terms appropriately_
`nlp('A Great Dane').nouns(0).json()[0]`
```
{
  "text": "Great Dane",
  "normal": "great dane",
  "terms": [
    {
      "text": "Great",
      "tags": [
        "ProperNoun",
        "Noun",
        "Singular",
        "Person"
      ],
      "pre": "",
      "post": " "
    },
    {
      "text": "Dane",
      "tags": [
        "MaleName",
        "FirstName",
        "Person",
        "ProperNoun",
        "Singular",
        "Noun"
      ],
      "pre": "",
      "post": ""
    }
  ],
  "article": ""
}
```

**Reply by spencermountain** [reply]
yep, both will work in v14
cheers

**Reply by scpedicini** [reply]
Hey Spencer,

Sorry to bug you again. (At this point I should be paying you for support). Earlier you mentioned that you'd be removing the list[0] access, but I'm having trouble finding a way to access the reduced form of a possessive noun or a plural noun using the new json(). I used to use the 'root' field.

For example:
`nlp("the king's toy").nouns(0).json()[0]`

Breaks down into the following json:
```
{
  "text": "king's toy",
  "normal": "king's toy",
  "terms": [
    {
      "text": "king's",
      "tags": [
        "Honorific",
        "Noun",
        "Possessive",
        "Person",
        "ProperNoun",
        "Singular"
      ],
      "pre": "",
      "post": " "
    },
    {
      "text": "toy",
      "tags": [
        "Person",
        "ProperNoun",
        "Singular",
        "Noun",
        "MaleName",
        "FirstName"
      ],
      "pre": "",
      "post": ""
    }
  ],
  "article": ""
}
```

But if I make use of 
`nlp(`the king's toy`).nouns(0).list[0].terms()`

I have the "reduced" form which gives me "king".

```
[
  {
    "text": "king's",
    "clean": "king's",
    "reduced": "king",
    "root": null,
    "implicit": null,
    "pre": "",
    "post": " ",
    "tags": {
      "Honorific": true,
      "Noun": true,
      "Possessive": true,
      "Person": true,
      "ProperNoun": true,
      "Singular": true
    },
    "prev": "the-8aNDO0M",
    "next": "toy-H1JiiXu",
    "id": "king's-0ug4b8D",
    "isA": "Term"
  },
  {
    "text": "toy",
    "clean": "toy",
    "reduced": "toy",
    "root": null,
    "implicit": null,
    "pre": "",
    "post": "",
    "tags": {
      "Person": true,
      "ProperNoun": true,
      "Singular": true,
      "Noun": true,
      "MaleName": true,
      "FirstName": true
    },
    "prev": "king's-0ug4b8D",
    "next": null,
    "id": "toy-H1JiiXu",
    "isA": "Term"
  }
]
```

Additionally, neither command (list or json) produces a singular version of a plural noun. For example nlp(`the kings of England`) I couldn't find a way to find the singular form. From a non-destructive standpoint, is there anything I can do besides spinning up another instance of nlp and calling toSingular()?

One thing I'm having to be aware of is performance, this NLP system is running against potentially millions of sentences which is why I'm wary of having to spin up new instances of the nlp() object since I think it might cost me over time.

**Reply by spencermountain** [reply]
no prob. This is something I've struggled with too.

sometimes I do `.clone().nouns().toSingular()`

but there's also [.cache()](https://observablehq.com/@spencermountain/compromise-cache): 
https://runkit.com/spencermountain/60ef32911f1137001a28fd0c
That's how i'd do it. It's not perfect.
may be worth trying either, to see if it's faster to clone or run the cache.

i thought `.json({terms:{root:true}})` would have worked, but maybe it's a bug.

once you've run the cache, you should be able to do
```js
doc.match('~king~') //matches kings
```

agreed that this is all awkward right now, and I'd love to clean it up.

**Reply by spencermountain** [reply]
yeah, sometimes I think it's easier to just loop through json, and have people do custom stuff, but then you always end up having to re-connect back to the original document, like carry around indexes or offsets or something, which is hell.

it's frustrating that a compromise match result doesn't play as nicely as json - and feels so frameworky. I thought adding .map and .foreach would help with that, but it doesn't really feel like it.

**Reply by scpedicini** [reply]
thanks, yeah I get the frustation, I worked on a fairly substantial library where we had four or five different means of exposing the same data, but it just ended up getting really confusing and half the time you'd end up mixing the different accessor systems together which would get even more dicey. 

Part of my problem is that I would guess that my use-case is somewhat unusual for compromise, in so far as I want to extract a noun(X) value and then use the tags on each term *within* the noun and reduce each term to a singular form, for example compromise would consider this `the king's brother's dog's toys` as a single noun, and from there I had formerly in v11 reduced each term using the list[0] and 'root' form that was automagically included.

But I think your suggestion of using toSingular() might solve this for me:
`nlp("the king's brother's dog's toys").nouns().clone().nouns().toSingular().json()[0]
`
Unfortunately, it looks like it strips out original casing. Is that intentional?

`nlp("Those are Great Danes").nouns(0).json()[0]`
Each term "text" field contain original capitalized case "Great" and "Danes".

```
nlp("Those are Great Danes").nouns(0).clone().nouns(0).toSingular().json()[0]
Terms[0] => "Great"
Terms[1] => "dane" ; has been lowercased
```

**Reply by spencermountain** [reply]
that looks like a bug, please file it
thanks

**Reply by scpedicini** [reply]
Will do~
https://github.com/spencermountain/compromise/issues/855
